src/WebUI/Pages/Logout.cshtml.cs

```diff
@@ -0,0 +1,21 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.Extensions.Logging;
+
+namespace Codidact.WebUI.Pages
+{
+    public class LogoutModel : PageModel
+    {
+        private readonly ILogger<IndexModel> _logger;
+
+        public LogoutModel(ILogger<IndexModel> logger)
+        {
+            _logger = logger;
+        }
+
+        public IActionResult OnGet()
```

Comment:
```suggestion
        public IActionResult OnPost()
```

I wasn't able to verify this because the authentication server didn't want to startup, but this looks like a CSRF vulnerability. Razor Pages will automatically verify CRSF tokens but that only happens for post requests.

Reply:
https://docs.microsoft.com/en-us/aspnet/core/razor-pages/?view=aspnetcore-3.1&tabs=visual-studio#xsrf
https://security.stackexchange.com/a/115810
Seems fine to me

Reply:
> GET endpoints should be idempotent - **actions that enact a change and do not simply retrieve data should require sending a POST (or other HTTP method) request**. POST endpoints should not interchangeably accept GET requests with parameters in the query string. &mdash; [MDN: HTTP cookies - Cross-site request forgery (CSRF)][1]

A malicious site could include some Java Script in the background that logs you out of `codidact.com`. This issue can be resolved by validating an CSRF token, which Razor Pages will do automatically, but not for 'GET' requests. 

  [1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Cross-site_request_forgery_CSRF

Reply:
@asynts It is a somewhat major issue, because [it doesn't take a malicious user](https://thedailywtf.com/articles/The_Spider_of_Doom) to trigger changes with GET requests, and even if that won't happen in this specific instance, if we don't consciously make this a part of our coding practice, it'll trip us up elsewhere.

Reply:
The GET doesnt change state, it only creates a url to go to the auth instance.
Also take a look at the example by IS, they didn't add anti forgery validation:
https://github.com/IdentityServer/IdentityServer4/blob/master/samples/Quickstarts/6_AspNetIdentity/src/MvcClient/Controllers/HomeController.cs

Reply:
It's probably best to change this to use POST.

https://stackoverflow.com/questions/3521290/logout-get-or-post

Reply:
This is not a log out call, this is a redirect call. The actual logout is in auth. 
If there is no auth instance there is no logout.

Reply:
> it only creates a url to go to the auth instance

I tried digging through their code and this doesn't seem to be the case:

 1. [IdentityServerAuthenticationService.cs#89][2]
 2. [DefaultUserSession.cs#L210-L222][3]

It's always difficult to tell with all the dependency injection stuff, but it seems, that the logout is done by removing the authentication cookie without any request to the identity server.

From their own documentation:

  > Typically you should prompt the user for signout (meaning require a POST), otherwise an attacker could hotlink to your logout page causing the user to be automatically logged out. &mdash;&nbsp;[docs.identityserver.io - Sign-out - Removing the authentication cookie][1]

Seems like an oversight to me.

  [1]: http://docs.identityserver.io/en/latest/topics/signout.html#removing-the-authentication-cookie
  [2]: https://github.com/IdentityServer/IdentityServer4/blob/ef7c904733add9d5675da59d01b79f51774c1a2f/src/IdentityServer4/src/Hosting/IdentityServerAuthenticationService.cs#L88

  [3]: https://github.com/IdentityServer/IdentityServer4/blob/ef7c904733add9d5675da59d01b79f51774c1a2f/src/IdentityServer4/src/Services/Default/DefaultUserSession.cs#L210-L222

Reply:
> I tried digging through their code and this doesn't seem to be the case:

Signout is part of PageModel, not identity server. Its part of .net core app.
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.signout?view=aspnetcore-3.1

Reply:
Looks like you are right. It just redirects to the identity server stuff where the user is prompted if he wants to logout, clicking 'yes' would make a 'POST' request.

Reply:
`Signout("cookie")` _will_ clear the application’s cookie though. Yes, the signout on OIDC will cause a redirect to the authentication provider (to sign out _there_) but before that, it will immediately sign out of the local application. So yes, a GET can be harmful here.

Reply:
@poke Disregard the comment I made above, I later noticed my mistake and created codidact/authentication#5.

Reply:
@asynts poke was saying that when the Signout("cookie") is executed it deletes the cookie before it moves to the auth project. I couldn't verify this to be honest and I looked at the source code page above too and I didn't see this happening. Frankly I am not sure and this needs further testing. To test need to just redirect and not logout and see if cookie is still there.

Reply:
That is correct. A GET on an action that will call `Signout("cookie")` will _immediately_ clear the application cookie in the HTTP response. If you pair that with a `Signout("oidc")`, then the _same_ HTTP response will _also_ contain a redirect to the remote authentication provider’s signout.

So the GET will have an effect, regardless of whether the user would follow that redirect or not.

If you don’t trust my word on this: the sign-out result [signs out each scheme](https://github.com/dotnet/aspnetcore/blob/v3.1.2/src/Mvc/Mvc.Core/src/SignOutResult.cs#L107-L110) which will eventually call [`SignOutAsync` on each authentication handler](https://github.com/dotnet/aspnetcore/blob/v3.1.2/src/Http/Authentication.Core/src/AuthenticationService.cs#L225). While the OIDC scheme will create [a redirect](https://github.com/dotnet/aspnetcore/blob/v3.1.2/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L243-L249), the cookie scheme [will also clear the cookie](https://github.com/dotnet/aspnetcore/blob/v3.1.2/src/Security/Authentication/Cookies/src/CookieAuthenticationHandler.cs#L363-L366). And that happens both in the same response.

Reply:
Can we complete this discussion in order to be able to finish reviewing/merge this PR soon?

Reply:
I would not rush this, it should be carefully verified and fixed. This would mean setting up a testing environment, maybe we can move it to a different issue and merge this now?

Reply:
I have created #67 for this, consider this "resolved" and just merge it for now.

Reply:
Switched to POST for logout